src/mpid/ch3/channels/nemesis/netmod/ofi/ofi_init.c

```diff
@@ -22,6 +22,9 @@ static inline int compile_time_checking();
         If non-null, choose an OFI provider by name. If using with the CH4
         device and using an older libfabric installation than the recommended
         version to accompany this MPICH version, unexpected results may occur.
+        The OFI library itself provides a similar environment variable FI_PROVIDER.
+        If that environment variable is set, it will take priority over this
+        one to avoid crashing when MPICH and OFI expect different providers.
```

Comment:
I don't think this commit is necessary anymore.

For --
1. `FI_PROVIDER` controls libfabric behavior and `MPIR_CVAR_OFI_USE_PROVIDER` controls mpich behavior. Think about the environment as a global variable space. It is not a good practice to cross access variables from different code-space. It might be convenient sometimes, but it makes code more difficult to maintain nevertheless.

2. I think the main branch already achieves what you desire here. Reference https://github.com/pmodels/mpich/blob/231904221f2801fd9243e9704720c196368e9d62/src/mpid/ch4/netmod/ofi/init_provider.c#L125-L127
When the automatic selection selects a particular provider, which can be controlled with `FI_PROVIDER`, we'll use that selected provider name to initialize settings, thus achieve the same effect as we use MPIR_CVAR_OFI_USE_PROVIDER to pick that provider in the first place.

Reply:
The problem that we saw was that we would do something like set `MPIR_CVAR_OFI_USE_PROVIDER` in a module file, but a user would see OFI docs (or be familiar enough with OFI) and set `FI_PROVIDER` to something else (e.g., we set `sockets`, but they set `tcp;ofi_rxm`). This led to a mismatch, which caused a crash.

MPICH can check for `FI_PROVIDER` because it is higher in the stack than OFI, but not the other way around. So it made sense for us to try to see if the user had picked something.

Reply:
>  set `MPIR_CVAR_OFI_USE_PROVIDER` in a module file

I would think this is the root of evil. Having these *debug* settings hidden in a system file often spawns deep frustrations to users -- as in the case you described.

For a given system, if only certain providers is preferred, the system can simply configure and build it correctly, right?

Anyway, at this point, I think we can actually remove the use of `MPIR_CVAR_OFI_USE_PROVIDER` from ch4/ofi altogether. FI_PROVIDER can fully replace all the intended use-cases for `MPIR_CVAR_OFI_USE_PROVIDER` now.

Reply:
> Anyway, at this point, I think we can actually remove the use of `MPIR_CVAR_OFI_USE_PROVIDER` from ch4/ofi altogether. FI_PROVIDER can fully replace all the intended use-cases for `MPIR_CVAR_OFI_USE_PROVIDER` now.

I think this might actually be the best solution. I'm not sure what benefit `MPIR_CVAR_OFI_USE_PROVIDER` is still giving.

Reply:
I thought about this some more and I'm not sure that removing the CVAR would actually let us do what we want to do.

This is not a valid assumption:

> For a given system, if only certain providers is preferred, the system can simply configure and build it correctly, right?

There are many times where the people who build the OFI library and the people who build the MPICH library are different. There are also cases where you might want to build support for more than one provider. In those circumstances, if you want to also provide an MPICH build that picks the correct default, you need a way to communicate that with everyone. `MPIR_CVAR_OFI_USE_PROVIDER` serves that purpose because if it set, we use that value. If it's not set, this PR lets us choose the default we'd prefer (this let's us provide a "sockets" debug build and a "psm2" debug build as two separate things for instance).

If we drop `MPIR_CVAR_OFI_USE_PROVIDER`, we can't accomplish the same thing with `FI_PROVIDER` because we can't control a "default" value for that environment variable.

Reply:
I think a valid alternative where we **would** eliminate `MPIR_CVAR_OFI_USE_PROVIDER` would break this statement:

> I would think this is the root of evil. Having these debug settings hidden in a system file often spawns deep frustrations to users -- as in the case you described.

I think it would also fulfill our needs to set `FI_PROVIDER` in a module file (i.e., you ask for the MPICH debug build with PSM2 support and we set `FI_PROVIDER=psm2` in a module file). If we did that, provider selection would work properly if the user doesn't overwrite this value because calling `fi_getinfo` will only return providers that match the environment variable. If the user **does** overwrite the variable, they'll get what they ask for.

Reply:
Right, if you already *decided* to set environment variables (including user preference) in a module file, then setting `FI_PROVIDER` instead of setting `MPIR_CVAR_OFI_USE_PROVIDER` is valid. Anyway, I think I have lost you -- what is the argument *against* removing the `MPIR_CVAR_OFI_USE_PROVIDER` altogether at this point?

Reply:
You didn't follow my random stream of consciousness? 😁

I think it's fine to do that and we would switch to using the OFI variable in our module files.

Reply:
👍  You can do that (remove MPIR_CVAR_OFI_USE_PROVIDER) in this PR. Hopefully, our CI should work (since it is just removing an extra variable). If you would prefer me to create a separate PR, just let me know.

Reply:
Ok. I've added these changes here. I'll run some tests this afternoon. Hopefully this takes things in the direction we were thinking though.